### ISSUE TITLE
The jitted function outputs a different result

### ISSUE THREAD

**Issue by HeavyCrab** [google/flax#4620]
I try to implement my own neural network and to use `nnx.jit` to speed up the forward call. However, the jitted function outputs a different result compared to the unjitted version.

### System information
- OS Platform and Distribution: Linux Ubuntu 20.04
- Flax, jax, jaxlib versions:
```
flax                      0.10.4                   pypi_0    pypi
jax                       0.5.2                    pypi_0    pypi
jax-cuda12-pjrt           0.5.1                    pypi_0    pypi
jax-cuda12-plugin         0.5.1                    pypi_0    pypi
jaxlib                    0.5.1                    pypi_0    pypi
optax                     0.2.4                    pypi_0    pypi
orbax-checkpoint          0.11.8                   pypi_0    pypi

```
- Python version: 3.11.11
- GPU model and memory: RTX 3090, 25.31G
- CUDA version: 12.8

### Problem you have encountered:

This code can reproduce my problem:
```
from flax import nnx
import jax
from jax import numpy as jnp
import numpy as np

class PositionalEncoding(nnx.Module):

    def __init__(self, d_hid, n_position=2):
        self.pos_table = self._get_sinusoid_encoding_table(n_position, d_hid)

    def _get_sinusoid_encoding_table(self, n_position, d_hid):
        """ Sinusoid position encoding table """

        def get_position_angle_vec(position):
            return [position / np.power(10000, 2 * (hid_j // 2) / d_hid) for hid_j in range(d_hid)]

        sinusoid_table = np.array([get_position_angle_vec(pos_i) for pos_i in range(n_position)])
        sinusoid_table[:, 0::2] = np.sin(sinusoid_table[:, 0::2])
        sinusoid_table[:, 1::2] = np.cos(sinusoid_table[:, 1::2])

        return nnx.Variable(jnp.array(sinusoid_table[None, ...]))

    def __call__(self, x):
        return x + self.pos_table[:, :x.shape[1]]

class mlp(nnx.Module):
    
    def __init__(self, in_dim, rngs):
        self.pos_embed = PositionalEncoding(d_hid=in_dim, n_position=30)
        
        self.in_dim = in_dim
        
        self.layers = nnx.Linear(in_features=in_dim, out_features=in_dim, rngs=rngs)
        
    def __call__(self, x):
        x = self.pos_embed(x)
        x = self.layers(x)
        return x



def forward(graphdef: nnx.GraphDef, state: nnx.State, x: jax.Array, y:jax.Array) -> tuple[jax.Array, nnx.State]:
    model = nnx.merge(graphdef, state)
    pred = model(x)
    loss = jnp.mean((pred - y) ** 2)
    _, state = nnx.split(model)
    return loss, state
jax_jitted_forward = jax.jit(forward)
nnx_jitted_forward = nnx.jit(forward)


in_dim = 32
rngs = nnx.Rngs(0)
model = mlp(in_dim, rngs)

key = jax.random.key(0)
batch_size, seq_len = 2, 4
x = jax.random.normal(key, (batch_size, seq_len, in_dim))
y = jax.random.normal(key, (batch_size, seq_len, in_dim))


model.eval()
pred = model(x)
loss = jnp.mean((pred - y) ** 2)
print(f'__call__ loss: {loss}')



graphdef, state = nnx.split(model)
loss, state = forward(graphdef, state, x=x, y=y)
nnx.update(model, state)
print(f'forward loss: {loss}')

graphdef, state = nnx.split(model)
loss, state = jax_jitted_forward(graphdef, state, x=x, y=y)
nnx.update(model, state)
print(f'jax jitted forward loss: {loss}')

graphdef, state = nnx.split(model)
loss, state = nnx_jitted_forward(graphdef, state, x=x, y=y)
nnx.update(model, state)
print(f'nnx jitted forward loss: {loss}')

pred = nnx.jit(model.__call__)(x)
loss = jnp.mean((pred - y) ** 2)
print(f'nnx jitted __call__ loss: {loss}')
```

I expect these different ways to calculate loss to be the same, however the jitted versions have different output compared to unjitted versions.

The output of the code above is

```
__call__ loss: 2.391187906265259
forward loss: 2.391187906265259
jax jitted forward loss: 2.3891987800598145
nnx jitted forward loss: 2.3891987800598145
nnx jitted __call__ loss: 2.3891987800598145
```

**Comment by selamw1** [comment]
Hi @HeavyCrab,
Thanks for the repro. However, the provided code doesn't reproduce the issue on either Colab with an NVIDIA T4 GPU or on an NVIDIA L4 GPU. Both GPUs print identical loss values, regardless of whether the functions are jitted or not. Here is a [gist](https://colab.sandbox.google.com/gist/selamw1/2324074f56a95c185b8c5b59dec7c158/4620.ipynb#) for your reference. 
Could you please double-check your JAX installation for GPU support or confirm if this issue is specific to RTX 3090 GPUs?

**Comment by cgarciae** [comment]
Not seeing a difference on my M1 either.

**Comment by HeavyCrab** [comment]
> Hi [@HeavyCrab](https://github.com/HeavyCrab), Thanks for the repro. However, the provided code doesn't reproduce the issue on either Colab with an NVIDIA T4 GPU or on an NVIDIA L4 GPU. Both GPUs print identical loss values, regardless of whether the functions are jitted or not. Here is a [gist](https://colab.sandbox.google.com/gist/selamw1/2324074f56a95c185b8c5b59dec7c158/4620.ipynb#) for your reference. Could you please double-check your JAX installation for GPU support or confirm if this issue is specific to RTX 3090 GPUs?

I cannot reproduce it on Colab T4 GPU either.
The package versions listed above are correct. But I wrote the CUDA version incorrectly (sorry for that). The correct CUDA version is 11.7.
```
GPU Device: NVIDIA GeForce RTX 3090
GPU Memory: 25.31 GB
CUDA Version: 11.7
```
For more information, my driver is NVIDIA-Linux-x86_64-570.86.16.

Thank you!

**Comment by HeavyCrab** [comment]
I tested my code on another A100 server, this issue is reproducible on the A100 server (with different outputs for unjitted functions)

```
$ python test.py 
__call__ loss: 2.391246795654297
forward loss: 2.391246795654297
jax jitted forward loss: 2.3891987800598145
nnx jitted forward loss: 2.3891987800598145
nnx jitted __call__ loss: 2.3891987800598145
```

System: Ubuntu 20.04.6 LTS
GPU Device: NVIDIA A100 80GB PCIe
GPU Memory: 85.10 GB

packages:
```
flax                      0.10.4                   pypi_0    pypi
jax                       0.5.2                    pypi_0    pypi
jax-cuda12-pjrt           0.5.1                    pypi_0    pypi
jax-cuda12-plugin         0.5.1                    pypi_0    pypi
jaxlib                    0.5.1                    pypi_0    pypi
optax                     0.2.4                    pypi_0    pypi
orbax-checkpoint          0.11.8                   pypi_0    pypi
```
Driver version: 560.35.03


Here is the full conda `environment.yml`.

```
channels:
  - conda-forge
  - defaults
  - https://repo.anaconda.com/pkgs/main
  - https://repo.anaconda.com/pkgs/r
dependencies:
  - _libgcc_mutex=0.1=main
  - _openmp_mutex=5.1=1_gnu
  - atk-1.0=2.36.0=ha1a6a79_0
  - boost-cpp=1.82.0=hdb19cb5_2
  - bzip2=1.0.8=h5eee18b_6
  - c-ares=1.19.1=h5eee18b_0
  - ca-certificates=2025.2.25=h06a4308_0
  - cairo=1.16.0=hb05425b_5
  - expat=2.6.4=h6a678d5_0
  - font-ttf-dejavu-sans-mono=2.37=hd3eb1b0_0
  - font-ttf-inconsolata=2.001=hcb22688_0
  - font-ttf-source-code-pro=2.030=hd3eb1b0_0
  - font-ttf-ubuntu=0.83=h8b1ccd4_0
  - fontconfig=2.14.1=h55d465d_3
  - fonts-anaconda=1=h8fa9717_0
  - fonts-conda-ecosystem=1=hd3eb1b0_0
  - freetype=2.12.1=h4a9f257_0
  - fribidi=1.0.10=h7b6447c_0
  - gdk-pixbuf=2.42.10=h5eee18b_1
  - giflib=5.2.2=h5eee18b_0
  - glib=2.78.4=h6a678d5_0
  - glib-tools=2.78.4=h6a678d5_0
  - gobject-introspection=1.78.1=py311h42194e9_2
  - gperftools=2.9.1=h33c463a_1
  - graphite2=1.3.14=h295c915_1
  - graphviz=2.50.0=h78213b7_2
  - gtk2=2.24.33=h27e1c3a_3
  - gts=0.7.6=hb67d8dd_3
  - harfbuzz=10.2.0=hf296adc_0
  - icu=73.1=h6a678d5_0
  - jpeg=9e=h5eee18b_3
  - krb5=1.20.1=h143b758_1
  - lcms2=2.16=hb9589c4_0
  - ld_impl_linux-64=2.40=h12ee557_0
  - lerc=4.0.0=h6a678d5_0
  - libboost=1.82.0=h109eef0_2
  - libcurl=8.12.1=hc9e6f67_0
  - libdeflate=1.22=h5eee18b_0
  - libedit=3.1.20230828=h5eee18b_0
  - libev=4.33=h7f8727e_1
  - libffi=3.4.4=h6a678d5_1
  - libgcc-ng=11.2.0=h1234567_1
  - libgd=2.3.3=h6a678d5_3
  - libglib=2.78.4=hdc74915_0
  - libgomp=11.2.0=h1234567_1
  - libiconv=1.16=h5eee18b_3
  - libnghttp2=1.57.0=h2d74bed_0
  - libpng=1.6.39=h5eee18b_0
  - librsvg=2.56.3=hf6914bd_1
  - libssh2=1.11.1=h251f7ec_0
  - libstdcxx-ng=11.2.0=h1234567_1
  - libtiff=4.5.1=hffd6297_1
  - libtool=2.4.7=h6a678d5_0
  - libunwind=1.6.2=h9c3ff4c_0
  - libuuid=1.41.5=h5eee18b_0
  - libwebp=1.2.4=h11a3e52_1
  - libwebp-base=1.2.4=h5eee18b_1
  - libxcb=1.15=h7f8727e_0
  - libxml2=2.13.5=hfdd30dd_0
  - lz4-c=1.9.4=h6a678d5_1
  - ncurses=6.4=h6a678d5_0
  - ninja=1.12.1=h06a4308_0
  - ninja-base=1.12.1=hdb19cb5_0
  - nspr=4.35=h6a678d5_0
  - nss=3.89.1=h6a678d5_0
  - openjpeg=2.5.2=he7f1fd0_0
  - openssl=3.0.16=h5eee18b_0
  - pango=1.50.7=h0fee60c_1
  - pcre2=10.42=hebb0a14_1
  - perl=5.38.2=0_h5eee18b_perl5
  - pip=25.0=py311h06a4308_0
  - pixman=0.40.0=h7f8727e_1
  - poppler=24.09.0=hcf11d46_1
  - poppler-data=0.4.11=h06a4308_1
  - python=3.11.11=he870216_0
  - readline=8.2=h5eee18b_0
  - setuptools=75.8.0=py311h06a4308_0
  - sqlite=3.45.3=h5eee18b_0
  - tk=8.6.14=h39e8969_0
  - wheel=0.45.1=py311h06a4308_0
  - xz=5.6.4=h5eee18b_1
  - zlib=1.2.13=h5eee18b_1
  - zstd=1.5.6=hc292b87_0
  - pip:
      - absl-py==2.1.0
      - aiosignal==1.3.2
      - attrs==25.3.0
      - cachetools==5.5.2
      - certifi==2025.1.31
      - charset-normalizer==3.4.1
      - chex==0.1.89
      - click==8.1.8
      - cloudpickle==3.1.1
      - cmake==3.31.6
      - contourpy==1.3.1
      - cycler==0.12.1
      - cython==3.0.11
      - docker-pycreds==0.4.0
      - etils==1.12.2
      - farama-notifications==0.0.4
      - filelock==3.18.0
      - flax==0.10.4
      - fonttools==4.56.0
      - frozenlist==1.5.0
      - fsspec==2025.3.0
      - gitdb==4.0.12
      - gitpython==3.1.44
      - google-auth==2.38.0
      - google-auth-oauthlib==1.0.0
      - grpcio==1.71.0
      - gymnasium==1.0.0
      - humanize==4.12.1
      - idna==3.10
      - importlib-resources==6.5.2
      - jax==0.5.2
      - jax-cuda12-pjrt==0.5.1
      - jax-cuda12-plugin==0.5.1
      - jaxlib==0.5.1
      - jinja2==3.1.6
      - jsonschema==4.23.0
      - jsonschema-specifications==2024.10.1
      - kiwisolver==1.4.8
      - kornia==0.6.6
      - lit==18.1.8
      - markdown==3.7
      - markdown-it-py==3.0.0
      - markupsafe==3.0.2
      - matplotlib==3.7.5
      - mdurl==0.1.2
      - ml-dtypes==0.5.1
      - mpmath==1.3.0
      - msgpack==1.1.0
      - nest-asyncio==1.6.0
      - networkx==3.4.2
      - numpy==1.26.4
      - nvidia-cublas-cu11==11.10.3.66
      - nvidia-cublas-cu12==12.8.4.1
      - nvidia-cuda-cupti-cu11==11.7.101
      - nvidia-cuda-cupti-cu12==12.8.90
      - nvidia-cuda-nvcc-cu12==12.8.93
      - nvidia-cuda-nvrtc-cu11==11.7.99
      - nvidia-cuda-runtime-cu11==11.7.99
      - nvidia-cuda-runtime-cu12==12.8.90
      - nvidia-cudnn-cu11==8.5.0.96
      - nvidia-cudnn-cu12==9.8.0.87
      - nvidia-cufft-cu11==10.9.0.58
      - nvidia-cufft-cu12==11.3.3.83
      - nvidia-curand-cu11==10.2.10.91
      - nvidia-cusolver-cu11==11.4.0.1
      - nvidia-cusolver-cu12==11.7.3.90
      - nvidia-cusparse-cu11==11.7.4.91
      - nvidia-cusparse-cu12==12.5.8.93
      - nvidia-nccl-cu11==2.14.3
      - nvidia-nccl-cu12==2.26.2
      - nvidia-nvjitlink-cu12==12.8.93
      - nvidia-nvtx-cu11==11.7.91
      - oauthlib==3.2.2
      - opencv-python==4.11.0.86
      - opt-einsum==3.4.0
      - optax==0.2.4
      - orbax-checkpoint==0.11.8
      - packaging==24.2
      - pandas==2.2.3
      - pathtools==0.1.2
      - pillow==11.1.0
      - promise==2.3
      - protobuf==4.25.6
      - psutil==7.0.0
      - pyasn1==0.6.1
      - pyasn1-modules==0.4.1
      - pygments==2.19.1
      - pyparsing==3.2.1
      - pysocks==1.7.1
      - python-dateutil==2.9.0.post0
      - pytz==2025.1
      - pyyaml==6.0.2
      - ray==2.10.0
      - referencing==0.36.2
      - requests==2.32.3
      - requests-oauthlib==2.0.0
      - rich==13.9.4
      - rpds-py==0.23.1
      - rsa==4.9
      - scipy==1.15.2
      - seaborn==0.13.2
      - sentry-sdk==2.22.0
      - setproctitle==1.3.5
      - shortuuid==1.0.13
      - simplejson==3.20.1
      - six==1.17.0
      - smmap==5.0.2
      - sympy==1.13.3
      - tensorboard==2.14.0
      - tensorboard-data-server==0.7.2
      - tensorstore==0.1.72
      - toolz==1.0.0
      - torch==2.0.1
      - torchaudio==2.0.2
      - torchvision==0.15.2
      - treescope==0.1.9
      - triton==2.0.0
      - typing-extensions==4.12.2
      - tzdata==2025.1
      - urllib3==2.3.0
      - wandb==0.13.5
      - werkzeug==3.1.3
      - zipp==3.21.0
```
